src/render.rs

```diff
@@ -0,0 +1,140 @@
+use matrix_sdk::events::room::{
+    encrypted::EncryptedEvent,
+    member::{MemberEvent, MembershipState},
+    message::{
+        AudioMessageEventContent, EmoteMessageEventContent,
+        FileMessageEventContent, ImageMessageEventContent, MessageEvent,
+        MessageEventContent, NoticeMessageEventContent,
+        TextMessageEventContent, VideoMessageEventContent,
+    },
+};
+
+/// This trait describes events that can be rendered in the weechat UI
+pub(crate) trait RenderableEvent {
+    /// Convert the event into a string that will be displayed in the UI.
+    /// The displayname is taken as a parameter since it cannot be calculated from the event
+    /// context alone.
+    fn render(&self, displayname: &str) -> String;
+}
+
+impl RenderableEvent for EncryptedEvent {
+    // TODO: this is not implemented yet
+    fn render(&self, displayname: &str) -> String {
+        format!("{}\t{}", displayname, "Unable to decrypt message")
+    }
+}
+
+impl RenderableEvent for MemberEvent {
+    fn render(&self, displayname: &str) -> String {
+        let operation = match self.content.membership {
```

Comment:
This isn't quite right, a `MembershipState::Join` will be present if a user joins or if a user changes the display name. You'll need to look at the `prev_content` and `content` to figure out the difference.

Luckily Ruma already has a convenience method for this, you can use [membership_change()](https://docs.rs/ruma-events/0.21.1/ruma_events/room/member/struct.MemberEvent.html#method.membership_change) on the event to get the correct change.

Reply:
Like this? e2d8e966d74a0dfa0bce6e40df1db18ca02bbebb

Reply:
Yes.

Reply:
Then let's move per-user colors to another issue/PR if you are OK with this.